### PR TITLE
HPCC-29528 LDAP Connection pool should not retry dead connections

### PR DESCRIPTION
**Pull request by RussWhitehead**
LDAP Connection pool objects should not try reconncting to LDAP AD hosts that have been blacklisted

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by wangkx — APPROVED** [review]
@RussWhitehead looks good.

**Comment by RussWhitehead** [comment]
@ghalliday Please merge

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 1]
```diff
@@ -1019,12 +1022,27 @@ class CLdapConnection : implements ILdapConnection, public CInterface
...
     {
+        if(!m_connected)
+            return connect();
+
+        //Ensure we are not using a rejected LDAP AD host
+        StringBuffer hostbuf;
+        m_ldapconfig->getLdapHost(hostbuf);
+        if (strcmp(hostbuf.str(), m_connectedHost.str()))
+        {
+            if (m_ld)//different host, disconnect from previous and try to reconnect to new one
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp:1033]: reuse: Could have put this code in an disconnect() function, and called from here and later in this function.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 2]
```diff
@@ -1019,12 +1022,27 @@ class CLdapConnection : implements ILdapConnection, public CInterface
 
     virtual bool validate()
     {
+        if(!m_connected)
+            return connect();
+
+        //Ensure we are not using a rejected LDAP AD host
+        StringBuffer hostbuf;
+        m_ldapconfig->getLdapHost(hostbuf);
+        if (strcmp(hostbuf.str(), m_connectedHost.str()))
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp:1031]: code is correct as it stands, but !streq() is clearer to read.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**